segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java

```diff
@@ -749,4 +763,80 @@ private void shutdownWhenStopped(Service component, String componentName) {
     }
 
     //endregion
+
+    //region DirectSegmentWrapper
+
+    /**
+     * Direct Segment Access implementation.
+     */
+    @RequiredArgsConstructor
+    private class DirectSegmentWrapper implements DirectSegmentAccess {
+        @Getter
+        private final long segmentId;
+
+        @Override
+        public CompletableFuture<Long> append(byte[] data, Collection<AttributeUpdate> attributeUpdates, Duration timeout) {
+            ensureRunning();
+            logRequest("append", this.segmentId, data.length);
+            StreamSegmentAppendOperation operation = new StreamSegmentAppendOperation(this.segmentId, data, attributeUpdates);
+            return processAttributeUpdaterOperation(operation, new TimeoutTimer(timeout))
+                    .thenApply(v -> operation.getStreamSegmentOffset());
+        }
+
+        @Override
+        public CompletableFuture<Void> updateAttributes(Collection<AttributeUpdate> attributeUpdates, Duration timeout) {
+            ensureRunning();
+            logRequest("updateAttributes", this.segmentId, attributeUpdates);
+            UpdateAttributesOperation operation = new UpdateAttributesOperation(this.segmentId, attributeUpdates);
```

Comment:
This is an obvious observation, and I'm sure you thought about it, so I want to know what you have to say. For the methods implemented in this wrapper, there is some duplication of code with respect to the implementation of the same method in the encapsulating class. Is there an opportunity to remove such duplication? I'm putting this comment here, but it is valid for pretty much any method in this wrapper.

Reply:
Good observation. I consolidated most of them.